### PR TITLE
Support Legends

### PR DESCRIPTION
**Pull request by greimel**
```julia
tbl1 = (x = rand(100), y = rand(100), z = 20 .* rand(100), grp1 = rand(["a", "b"], 100), grp2 = rand(["c", "d"], 100))

cols = mapping(:x => "The X", :y => "The Y");
grp = mapping(marker = :grp1 => nonnumeric => "Group", color = :grp2 => "Other", markersize = :z);
scat = visual(Scatter)
aog = data(tbl1) * cols * scat * grp

fig = aog |> draw
Legend(fig.figure[1,2], aog)
fig
```

image

## Todo

- [x] combine legends when the same variable is used more than once
- [ ] <s> support Colorbar </s> (future PR)
- [ ] <s> automatically add legend when calling `draw(...)` (do you want that @piever?) </s>

### CODE REVIEW

**Comment by piever** [comment]
Thanks a lot, this looks pretty nice already!

I think there is no need to have the legend drawn automatically in `draw` (it'd be easy to add that later in any case), but it would help to have an example of drawing the legend in the Example Gallery (could probably be the same as the one you're showing above, plus one with a colorbar).

Btw, a couple of practical things to add docs to gallery.jl:
- to try things locally, you need to be on the master branch of documenter (otherwise, it does not handle many svgs in one page well)
- it's probably better to rebase this branch (conflict-free) before adding docs, because I've just modified the whole doc setup

**Comment by greimel** [comment]
It turned out that I need to do some adjustments for the more general setup here.

I've not pushed this yet, but locally this works

image

That is, you can mix and match `PlotType`s and continous/categorical scales now and use a variable more than once.

**Comment by greimel** [comment]
## What needs to be fixed future PRs

- [ ] correctly handle composite plot types (now fall back to `poly`)
- [ ] specifying the order of legend elements (should be poly then line then marker)
- [ ] correctly handle the case when no plottype is given in `visual` (currently this errors)

## A few examples

image

image

image

image

**Review by piever — CHANGES_REQUESTED** [review]

**Comment by piever** [comment]
Thanks for adding the consolidation bit. I've left some comments. Other than some formatting nitpicking, I think the main point is to have a clean structure for `SubLegend` and `SubLegendEntry` that allows for merging. My suggestion is something along the lines of

```julia
struct SubLegendEntry
    label::String
    visuals::Vector{Visual} # each plottype has its one attributes
end

struct SubLegend
    title::String
    entries::Vector{SubLegendEntry}
end
```

We should make sure that this allow for a `merge!` method on sublegends that combines entries in the correct way.

Then, consolidation becomes much easier using dicts (well, an `OrderedCollections.LittleDict` to remember the order). For example:

```julia
sublegend_dict = LittleDict{String, SubLegend}()
for sublegend in sublegends
    key = sublegend.title
    current = get!(sublegend_dict, key, Sublegend(key, SubLegendEntry[]))
    merge!(current, sublegend)
end
```

File: src/guides/legend.jl [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+		map(key2Ps[key]) do P
+			_legend(P, key, scale, title)
+		end
+	end |> StructArray
+	
+	consolidate_legends(out)
+end
+
+# ╔═╡ 2bcbfb12-47e5-471a-b78f-4dfedaae3a0d
+function _legend(P, attribute, scale::ContinuousScale, title)
```
[the remark this answers is not in the document]
Comment (piever) [reply src/guides/legend.jl:79]: (Thinking more carefully, scratch the above type hierarchy, I try to explain my idea more clearly in https://github.com/piever/SplitApplyPlot.jl/pull/32#issuecomment-830005943.)

**Comment by greimel** [comment]
Hi, I'll probably take a few days do address all this. I wanted to push a working version before the weekend.

**Comment by piever** [comment]
@greimel , just a heads up that I'm about to do some `Entries / AxisEntries` rework, so that some details of how things are done (in particular legend consolidations) will likely need to change. I was thinking to merge this as is (even if a bit preliminar) and adjust it as I do the refactor. (Otherwise you'd have to work on things that may likely need to change in the near future...)

Do you have some local changes you'd like to push or can I go ahead and merge this version?

**Comment by greimel** [comment]
hi @piever, sorry that this has stalled a bit. I've not worked on this in the meantime, I was busier than I thought with other things.

Despite the messy state it should be fine to merge. If necessary I will help cleaning this up later.

**Comment by greimel** [comment]
also feel free to commit the suggested changes.

**Pull request merged**